wicketd/src/mgs_manager.rs

```diff
@@ -0,0 +1,155 @@
+
+const MGS_POLL_INTERVAL: Duration = Duration::from_secs(10);
+const MGS_TIMEOUT: u32 = 3000; // 3 sec
+
+// 32 sleds, 2 switches, 2 PSCs (at some point)
+const MAX_COMPONENTS: Option<NonZeroU32> = NonZeroU32::new(36);
+
+// We support:
+//   * One outstanding query request from wicket
+//   * One outstanding update/interaction request from wicket
+//   * Room for some timeouts and re-requests from wicket.
+const CHANNEL_CAPACITY: usize = 8;
+
+/// Channel errors result only from system shutdown. We have to report them to
+/// satisfy function calls, so we define an error type here.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ShutdownInProgress;
+
+#[derive(Debug)]
+pub enum MgsRequest {
+    GetInventory {
+        etag: Option<String>,
+        reply_tx: oneshot::Sender<Arc<RackV1Inventory>>,
+    },
+}
+
+/// A mechanism for interacting with the  MgsManager
+pub struct MgsHandle {
+    tx: tokio::sync::mpsc::Sender<MgsRequest>,
+}
+
+impl MgsHandle {
+    pub async fn get_inventory(
+        &self,
+    ) -> Result<Arc<RackV1Inventory>, ShutdownInProgress> {
+        let (reply_tx, reply_rx) = oneshot::channel();
+        let etag = None;
+        self.tx
+            .send(MgsRequest::GetInventory { etag, reply_tx })
+            .await
+            .map_err(|_| ShutdownInProgress)?;
+        reply_rx.await.map_err(|_| ShutdownInProgress)
+    }
+}
+
+/// The entity responsible for interacting with MGS
+///
+/// `MgsManager` will poll MGS periodically to update its local information,
+/// and wicket will poll MGS periodically for updates. Inventory/status
+/// requests (e.g. HTTP GET requests to wicketd) will be served from the
+/// MgsManager and not result in a new underlying request to MGS. This keeps
+/// interaction to a constant amount of work, and limits bursts to any updates
+/// that require further pulling from MGS.
+///
+/// Update interacions from Wicket (e.g. HTTP POST/PUT requests to wicketd)
+/// will be forwarded one at a time to MGS. We only allow one outstanding
+/// update at a time. If the update request is to actually update software,
+/// and there is a long running update in progress, the request will not
+/// be forwarded and a relevant response will be returned to wicket.
+pub struct MgsManager {
+    log: Logger,
+    tx: mpsc::Sender<MgsRequest>,
+    rx: mpsc::Receiver<MgsRequest>,
+    mgs_client: gateway_client::Client,
+    // Remove the need to copy a potentially sizeable amount of data
+    inventory: Arc<RackV1Inventory>,
+}
+
+impl MgsManager {
+    pub fn new(log: &Logger, mgs_addr: SocketAddrV6) -> MgsManager {
+        let log = log.new(o!("component" => "wicketd MgsManager"));
+        let (tx, rx) = mpsc::channel(CHANNEL_CAPACITY);
+        let endpoint =
+            format!("http://[{}]:{}", mgs_addr.ip(), mgs_addr.port());
+        info!(log, "MGS Endpoint: {}", endpoint);
+        let mgs_client = gateway_client::Client::new(&endpoint, log.clone());
```

Comment:
Not suggesting a change here necessarily, but just noting from past experience - progenitor clients created via `::new()` implicitly have a [15 second timeout](https://github.com/oxidecomputer/progenitor/blob/4e2dcc5365ea4282af7b1851c5ae96305cb51d9a/progenitor-impl/src/lib.rs#L247-L253) on requests. If that's not what we want, we need to use `new_with_client()` to override it.

Reply:
Thanks for the heads up!